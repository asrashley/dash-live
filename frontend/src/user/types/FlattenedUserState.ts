import type { InitialUserState } from "../types/InitialUserState";


export type FlattenedUserState = Omit<InitialUserState, 'groups'> & {
  adminGroup: boolean;
  mediaGroup: boolean;
  userGroup: boolean;
};
