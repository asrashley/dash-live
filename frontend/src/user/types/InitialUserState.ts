export type InitialUserState = {
    pk?: number;
    username?: string;
    email?: string;
    lastLogin: string | null;
    mustChange: boolean;
    groups: string[];
};
