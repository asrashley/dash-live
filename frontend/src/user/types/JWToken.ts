export type JWToken = {
  expires: string;
  jwt: string;
};
