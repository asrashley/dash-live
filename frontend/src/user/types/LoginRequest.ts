export interface LoginRequest {
    username: string;
    password: string;
    rememberme: boolean;
}