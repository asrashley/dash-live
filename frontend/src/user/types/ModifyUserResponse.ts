import type { InitialUserState } from "./InitialUserState";

export interface ModifyUserResponse {
    success: boolean;
    errors: string[];
    user?: InitialUserState;
}
