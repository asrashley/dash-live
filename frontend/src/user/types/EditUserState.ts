import { FlattenedUserState } from "./FlattenedUserState";


export type EditUserState = FlattenedUserState & {
  password?: string;
  confirmPassword?: string;
};
