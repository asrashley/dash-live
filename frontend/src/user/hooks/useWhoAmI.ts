import { createContext } from "preact";
import { useCallback } from "preact/hooks";
import { type ReadonlySignal, useComputed, useSignal } from "@preact/signals";

import { UserState } from "../types/UserState";
import { InitialUserState } from "../types/InitialUserState";
import { FlattenedUserState } from "../types/FlattenedUserState";
import { flattenUserState } from "../utils/flattenUserState";

export interface UseWhoAmIHook {
  user: ReadonlySignal<UserState>;
  flattened: ReadonlySignal<FlattenedUserState>;
  setUser: (ius: InitialUserState | null) => void;
}

export const WhoAmIContext = createContext<UseWhoAmIHook>(null);

const blankState: InitialUserState = {
  mustChange: false,
  lastLogin: null,
  groups:[],
};

export function useWhoAmI(): UseWhoAmIHook {
  const userInfo = useSignal<InitialUserState>(structuredClone(blankState));
  const user = useComputed<UserState>(() => ({
    ...userInfo.value,
    isAuthenticated: userInfo.value.pk !== undefined,
    permissions: {
      admin: userInfo.value.groups.includes('ADMIN'),
      media: userInfo.value.groups.includes('MEDIA'),
      user: userInfo.value.groups.includes('USER'),
    },
  }));
  const flattened = useComputed<FlattenedUserState>(() => flattenUserState(user.value));

  const setUser = useCallback((ius: InitialUserState | null) => {
    if (ius) {
      userInfo.value = structuredClone(ius);
    } else {
      userInfo.value = structuredClone(blankState);
    }
  }, [userInfo]);

  return { user, flattened, setUser };
}
