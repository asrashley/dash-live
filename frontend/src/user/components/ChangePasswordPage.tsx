import { useCallback, useContext } from "preact/hooks";
import { useComputed, useSignal } from "@preact/signals";
import { useLocation } from "wouter-preact";

import { uiRouteMap } from "@dashlive/routemap";

import { UserValidationErrors } from "../../hooks/useAllUsers";
import { EditUserCard } from "./EditUserCard";
import { EndpointContext } from "../../endpoints";
import { useMessages } from "../../hooks/useMessages";
import { WhoAmIContext } from "../../hooks/useWhoAmI";
import { validateUserState } from "../utils/validateUserState";
import { FlattenedUserState } from "../../types/FlattenedUserState";
import { ProtectedPage } from "../../components/ProtectedPage";

const only = ['email', 'password', 'confirmPassword'];

const backUrl = uiRouteMap.home.url();

export function ChangePassword() {
  const [, setLocation] = useLocation();
  const apiRequests = useContext(EndpointContext);
  const { flattened, setUser } = useContext(WhoAmIContext);
  const { appendMessage } = useMessages();
  const changes = useSignal<Partial<FlattenedUserState>>({});
  const disabledFields = useSignal<Record<string, boolean>>({
    username: true,
    mustChange: true,
    groups: true,
  });
  const user = useComputed(() => {
    const usr: FlattenedUserState = {
      adminGroup: false,
      mediaGroup: false,
      userGroup: false,
      mustChange: true,
      lastLogin: null,
      ...flattened.value,
      ...changes.value,
    };
    return usr;
  });
  const errors = useComputed<UserValidationErrors>(() => validateUserState(user.value));

  const setValue = useCallback(
    (field: string, value: string | number | boolean) => {
      changes.value = {
        ...changes.value,
        [field]: value,
      };
    },
    [changes]
  );

  const saveChanges = useCallback(async () => {
    try {
      const result = await apiRequests.editUser(user.value);
      if (result.success) {
        appendMessage("success", "Password successfully modified");
        setUser(result.user);
        setLocation(uiRouteMap.listUsers.url());
      } else {
        appendMessage("warning", "Failed to change password");
      }
    } catch(err) {
      appendMessage("warning", `Failed to change password - ${err}`);
    }
  }, [apiRequests, appendMessage, setLocation, setUser, user.value]);

  return (
    <div className="content mb-4" style="max-width: 65rem">
      <EditUserCard
        backUrl={backUrl}
        header="Change my password"
        saveTitle="Change Password"
        user={user}
        validationErrors={errors}
        disabledFields={disabledFields}
        onSave={saveChanges}
        setValue={setValue}
        only={only}
      />
    </div>
  );
}

export default function ChangePasswordPage() {
  return <ProtectedPage><ChangePassword /></ProtectedPage>;
}
