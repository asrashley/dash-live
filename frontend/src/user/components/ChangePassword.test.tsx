import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import { mock } from "vitest-mock-extended";
import userEvent from "@testing-library/user-event";
import { useLocation } from "wouter-preact";

import { renderWithProviders } from "../../test/renderWithProviders";
import { ApiRequests, EndpointContext } from "../../endpoints";
import { useMessages, UseMessagesHook } from "../../hooks/useMessages";
import { normalUser, UserModel } from "../../test/MockServer";
import { ModifyUserResponse } from "../../types/ModifyUserResponse";
import { EditUserState } from "../../types/EditUserState";

import ChangePasswordPage, { ChangePassword } from "./ChangePasswordPage";

vi.mock("wouter-preact", async (importOriginal) => {
  return {
    ...(await importOriginal()),
    useLocation: vi.fn(),
  };
});

vi.mock("../../hooks/useMessages", async (importOriginal) => {
  return {
    ...(await importOriginal()),
    useMessages: vi.fn(),
  };
});

describe("ChangePassword component", () => {
  const useLocationMock = vi.mocked(useLocation);
  const useMessagesMock = vi.mocked(useMessages);
  const setLocation = vi.fn();
  const apiRequests = mock<ApiRequests>();
  const messagesMock = mock<UseMessagesHook>();
  const userInfo: UserModel = {
    pk: normalUser.pk,
    username: normalUser.username,
    email: normalUser.email,
    password: "",
    mustChange: false,
    lastLogin: "",
    groups: normalUser.groups,
    refreshToken: null,
    accessToken: null,
  };

  beforeEach(() => {
    useLocationMock.mockReturnValue(["/change-password", setLocation]);
    useMessagesMock.mockReturnValue(messagesMock);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test("ChangePassword component matches snapshot", () => {
    const { asFragment, getByText, getByLabelText } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <ChangePassword />
      </EndpointContext.Provider>,
      { userInfo }
    );
    getByText("Change my password");
    const emailInp = getByLabelText("Email:") as HTMLInputElement;
    expect(emailInp.value).toEqual(normalUser.email);
    expect(asFragment()).toMatchSnapshot();
  });

  test("ChangePasswordPage matches snapshot", async () => {
    const { asFragment, findByText } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <ChangePasswordPage />
      </EndpointContext.Provider>,
      { userInfo }
    );
    await findByText("Change my password");
    expect(asFragment()).toMatchSnapshot();
  });

  test("can change password", async () => {
    const user = userEvent.setup();
    const { getByLabelText, getByText } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <ChangePassword />
      </EndpointContext.Provider>,
      { userInfo }
    );
    const saveChanges = new Promise<EditUserState>((resolve) => {
      apiRequests.editUser.mockImplementation(async (user: EditUserState) => {
        const resp: ModifyUserResponse = {
          success: true,
          errors: [],
          user: {
            ...user,
            groups: [],
          },
        };
        resolve(user);
        return resp;
      });
    });
    const newPassword = "new!password!";
    const pwdInp = getByLabelText("Password:") as HTMLInputElement;
    const confirm = getByLabelText("Confirm Password:") as HTMLInputElement;
    await user.click(pwdInp);
    await user.clear(pwdInp);
    await user.type(pwdInp, newPassword);
    const saveBtn = getByText("Change Password") as HTMLButtonElement;
    expect(saveBtn.disabled).toEqual(true);
    await user.click(confirm);
    await user.clear(confirm);
    await user.type(confirm, newPassword);
    expect(saveBtn.disabled).toEqual(false);
    await user.click(saveBtn);
    await expect(saveChanges).resolves.toEqual(
      expect.objectContaining({
        pk: normalUser.pk,
        email: normalUser.email,
        password: newPassword,
        confirmPassword: newPassword,
      })
    );
    expect(messagesMock.appendMessage).toHaveBeenCalledWith(
      "success",
      "Password successfully modified"
    );
  });

  test("fails to change password", async () => {
    const user = userEvent.setup();
    const { getByLabelText, getByText } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <ChangePassword />
      </EndpointContext.Provider>,
      { userInfo }
    );
    const saveChanges = new Promise<EditUserState>((resolve) => {
      apiRequests.editUser.mockImplementation(async (user: EditUserState) => {
        const resp: ModifyUserResponse = {
          success: false,
          errors: ["server error"],
        };
        resolve(user);
        return resp;
      });
    });
    const newPassword = "new!password!";
    const pwdInp = getByLabelText("Password:") as HTMLInputElement;
    const confirm = getByLabelText("Confirm Password:") as HTMLInputElement;
    await user.click(pwdInp);
    await user.clear(pwdInp);
    await user.type(pwdInp, newPassword);
    const saveBtn = getByText("Change Password") as HTMLButtonElement;
    await user.click(confirm);
    await user.clear(confirm);
    await user.type(confirm, newPassword);
    expect(saveBtn.disabled).toEqual(false);
    await user.click(saveBtn);
    await expect(saveChanges).resolves.toEqual(
      expect.objectContaining({
        pk: normalUser.pk,
        email: normalUser.email,
        password: newPassword,
        confirmPassword: newPassword,
      })
    );
    expect(messagesMock.appendMessage).toHaveBeenCalledWith(
      "warning",
      "Failed to change password"
    );
  });

  test("can change email address", async () => {
    const user = userEvent.setup();
    const { getByLabelText, getByText } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <ChangePassword />
      </EndpointContext.Provider>,
      { userInfo }
    );
    const saveChanges = new Promise<EditUserState>((resolve) => {
      apiRequests.editUser.mockImplementation(async (user: EditUserState) => {
        const resp: ModifyUserResponse = {
          success: true,
          errors: [],
          user: {
            ...user,
            groups: [],
          },
        };
        resolve(user);
        return resp;
      });
    });
    const newEmail = "new.email@account.local";
    const email = getByLabelText("Email:") as HTMLInputElement;
    await user.click(email);
    await user.clear(email);
    await user.type(email, newEmail);
    const saveBtn = getByText("Change Password") as HTMLButtonElement;
    expect(saveBtn.disabled).toEqual(false);
    await user.click(saveBtn);
    await expect(saveChanges).resolves.toEqual(
      expect.objectContaining({
        pk: normalUser.pk,
        email: newEmail,
      })
    );
    expect(messagesMock.appendMessage).toHaveBeenCalledWith(
      "success",
      "Password successfully modified"
    );
  });
});
