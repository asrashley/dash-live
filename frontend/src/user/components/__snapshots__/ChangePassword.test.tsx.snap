// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChangePassword component > ChangePassword component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="content mb-4"
    style="max-width: 65rem;"
  >
    <div
      class="card "
      id="edit-user"
    >
      <div
        class="card-header"
      >
        Change my password
      </div>
      <div
        class="card-body"
      >
        <form>
          <div
            class="row mb-2 form-group "
          >
            <label
              class="col-2 col-form-label"
              for="model-email"
              id="label-email"
            >
              Email:
            </label>
            <div
              class="col-10"
            >
              <input
                aria-describedby="label-email"
                class="form-control is-valid "
                id="model-email"
                name="email"
                title="Email"
                type="email"
              />
            </div>
            <div
              class="invalid-feedback"
              style="display: none;"
            />
          </div>
          <div
            class="row mb-2 form-group "
          >
            <label
              class="col-2 col-form-label"
              for="model-password"
              id="label-password"
            >
              Password:
            </label>
            <div
              class="col-10"
            >
              <div
                class="input-group"
              >
                <input
                  aria-describedby="label-password"
                  class="form-control is-valid "
                  id="model-password"
                  name="password"
                  placeholder="***"
                  title="Password"
                  type="password"
                />
                <button
                  class="d-none"
                  type="button"
                >
                  Reveal
                </button>
              </div>
            </div>
            <div
              class="invalid-feedback"
              style="display: none;"
            />
          </div>
          <div
            class="row mb-2 form-group "
          >
            <label
              class="col-2 col-form-label"
              for="model-confirmPassword"
              id="label-confirmPassword"
            >
              Confirm Password:
            </label>
            <div
              class="col-10"
            >
              <div
                class="input-group"
              >
                <input
                  aria-describedby="label-confirmPassword"
                  class="form-control is-valid "
                  id="model-confirmPassword"
                  name="confirmPassword"
                  placeholder="***"
                  title="Confirm Password"
                  type="password"
                />
                <button
                  class="d-none"
                  type="button"
                >
                  Reveal
                </button>
              </div>
            </div>
            <div
              class="invalid-feedback"
              style="display: none;"
            />
          </div>
        </form>
        <div
          class="form-actions mt-2"
        >
          <button
            class="btn btn-primary me-3"
          >
            Change Password
          </button>
          <a
            class="btn btn-warning"
            href="/"
          >
            Cancel
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ChangePassword component > ChangePasswordPage matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="content mb-4"
    style="max-width: 65rem;"
  >
    <div
      class="card "
      id="edit-user"
    >
      <div
        class="card-header"
      >
        Change my password
      </div>
      <div
        class="card-body"
      >
        <form>
          <div
            class="row mb-2 form-group "
          >
            <label
              class="col-2 col-form-label"
              for="model-email"
              id="label-email"
            >
              Email:
            </label>
            <div
              class="col-10"
            >
              <input
                aria-describedby="label-email"
                class="form-control is-valid "
                id="model-email"
                name="email"
                title="Email"
                type="email"
              />
            </div>
            <div
              class="invalid-feedback"
              style="display: none;"
            />
          </div>
          <div
            class="row mb-2 form-group "
          >
            <label
              class="col-2 col-form-label"
              for="model-password"
              id="label-password"
            >
              Password:
            </label>
            <div
              class="col-10"
            >
              <div
                class="input-group"
              >
                <input
                  aria-describedby="label-password"
                  class="form-control is-valid "
                  id="model-password"
                  name="password"
                  placeholder="***"
                  title="Password"
                  type="password"
                />
                <button
                  class="d-none"
                  type="button"
                >
                  Reveal
                </button>
              </div>
            </div>
            <div
              class="invalid-feedback"
              style="display: none;"
            />
          </div>
          <div
            class="row mb-2 form-group "
          >
            <label
              class="col-2 col-form-label"
              for="model-confirmPassword"
              id="label-confirmPassword"
            >
              Confirm Password:
            </label>
            <div
              class="col-10"
            >
              <div
                class="input-group"
              >
                <input
                  aria-describedby="label-confirmPassword"
                  class="form-control is-valid "
                  id="model-confirmPassword"
                  name="confirmPassword"
                  placeholder="***"
                  title="Confirm Password"
                  type="password"
                />
                <button
                  class="d-none"
                  type="button"
                >
                  Reveal
                </button>
              </div>
            </div>
            <div
              class="invalid-feedback"
              style="display: none;"
            />
          </div>
        </form>
        <div
          class="form-actions mt-2"
        >
          <button
            class="btn btn-primary me-3"
          >
            Change Password
          </button>
          <a
            class="btn btn-warning"
            href="/"
          >
            Cancel
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
